### PR TITLE
Fix unrepresentable date crash in edit time entry view

### LINKED ISSUE #5015: Un-representable DateTime in EditTimeEntryViewController
### 👥 Impact
11 reports from 11 users in the last 7 days

### 🔗 Context
Might be related to #5014 

### 🛰 Link
https://appcenter.ms/orgs/Toggl/apps/Toggl-iOS/crashes/errors/4146125805u/overview

### 📝  Exception & Stack Trace
`Exception type and message`
```cs
DateTime.op_Addition (System.DateTime d, System.TimeSpan t)
TimeZoneInfo.ConvertTime (System.DateTimeOffset dateTimeOffset, System.TimeZoneInfo destinationTimeZone)
DateTimeToFormattedString.getDateTimeOffsetInCorrectTimeZone (System.DateTimeOffset value, System.TimeZoneInfo timeZone)
DateTimeToFormattedString.Convert (System.DateTimeOffset date, System.String format, System.TimeZoneInfo timeZoneInfo)
EditTimeEntryViewController+<>c.<ViewDidLoad>b__8_10 (System.DateTimeOffset stopTime, Toggl.Core.Models.Interfaces.IThreadSafePreferences preferences)
WithLatestFrom`3+_+FirstObserver[TFirst,TSecond,TResult].OnNext (TFirst value)
ExceptionServicesImpl.Rethrow (System.Exception exception)
ExceptionHelpers.Throw (System.Exception exception)
<.cctor>b__2_1 (System.Exception ex)
AnonymousSafeObserver`1[T].OnError (System.Exception error)
Sink`1[TTarget].ForwardOnError (System.Exception error)
(wrapper unknown) System.Object.gsharedvt_out()
WithLatestFrom`3+_+FirstObserver[TFirst,TSecond,TResult].OnNext (TFirst value)
(wrapper unknown) System.Object.gsharedvt_in()
(wrapper unknown) System.Object.gsharedvt_out()
Sink`1[TTarget].ForwardOnNext (TTarget value)
(wrapper unknown) System.Object.gsharedvt_in()
(wrapper unknown) System.Object.gsharedvt_out()
Select`2+Selector+_[TSource,TResult].OnNext (TSource value)
(wrapper unknown) System.Object.gsharedvt_in()
(wrapper unknown) System.Object.gsharedvt_out()
Sink`1[TTarget].ForwardOnNext (TTarget value)
(wrapper unknown) System.Object.gsharedvt_in()
(wrapper unknown) System.Object.gsharedvt_out()
Where`1+Predicate+_[TSource].OnNext (TSource value)
(wrapper unknown) System.Object.gsharedvt_in()
(wrapper unknown) System.Object.gsharedvt_out()
Sink`1[TTarget].ForwardOnNext (TTarget value)
(wrapper unknown) System.Object.gsharedvt_in()
(wrapper unknown) System.Object.gsharedvt_out()
ObserveOnObserverNew`1[T].DrainStep (System.Collections.Concurrent.ConcurrentQueue`1[T] q, System.Boolean delayError)
(wrapper unknown) System.Object.gsharedvt_in()
(wrapper unknown) System.Object.gsharedvt_out()
ObserveOnObserverNew`1[T].DrainShortRunning (System.Reactive.Concurrency.IScheduler recursiveScheduler)
(wrapper unknown) System.Object.gsharedvt_in()
(wrapper unknown) System.Object.gsharedvt_out()
<.cctor>b__20_1 (System.Reactive.Concurrency.IScheduler scheduler, System.Reactive.ObserveOnObserverNew`1[T] self)
NSRunloopScheduler+<>c__DisplayClass2_0`1[TState].<Schedule>b__0 ()
DispatchQueue.static_dispatcher_to_managed (System.IntPtr context)
(wrapper native-to-managed) CoreFoundation.DispatchQueue.static_dispatcher_to_managed(intptr)
(wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate)
UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName)
Application.Main (System.String[] args)
(wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate)
UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName)
Application.Main (System.String[] args)
```

### PR DESCRIPTION
**Pull request by arturs105**
## What's this?
This is a bit different than the other unrepresentable date crash. This one is not a result of adding a `TimeSpan` to a `DateTimeOffset`, but a result of calling `TimeZoneInfo.ConvertTime`.

### Relationships
Closes #5015

## Why do we want this?
💥➡️🚫💥 

## How is it done?
In the case of encountering unrepresentable date exception, just return the same `DateTimeOffset` that was passed in. This will probably lead to incorrect data being displayed in the app, but it won't crash. I also added some analytics, so we can find out, which `DateTimeOffset` and `TimeZoneInfo` combinations cause the crash.

### Why not another way?
Can't really come up with anything else, since we have no idea, how, why and what data causes these exceptions.

The docs for `TimeZoneInfo.ConvertTime` say that it throws this exception when "The value of the `dateTime` parameter represents an invalid time", which means that we are passing in the unrepresentable `DateTimeOffset` and that date time might be created by #5014.

### Side effects
The app will show the unrepresentable date time instead of crashing.

## :squid: Permissions
🦑

### CODE REVIEW

File: Toggl.Core.UI/Transformations/DateTimeToFormattedString.cs [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
 namespace Toggl.Core.UI.Transformations
 {
     public class DateTimeToFormattedString
     {
-        public static string Convert(DateTimeOffset date, string format, TimeZoneInfo timeZoneInfo = null)
+        public static string Convert(
+            DateTimeOffset date,
+            string format,
+            TimeZoneInfo timeZoneInfo = null,
+            IAnalyticsService analyticsService = null)
```
Comment (kornelijepetak) [inline Toggl.Core.UI/Transformations/DateTimeToFormattedString.cs]: You don't seem to be using the `analyticsService` in this method. Why the parameter then?

Comment (arturs105) [reply Toggl.Core.UI/Transformations/DateTimeToFormattedString.cs]: 🤦‍♂ it was meant to be passed to `getDateTimeOffsetInCorrectTimeZone`

**Comment by amulware** [comment]
I definitely do not like the idea of catching exceptions if that leads to bad state. Can't we re-throw an exception with the additional information to help us debug the problem?

Also, why are we using analytics events instead of the app center error api?

File: Toggl.Core.UI/Transformations/DateTimeToFormattedString.cs [hunk 2]
```diff
@@ -1,21 +1,42 @@
...
+            TimeZoneInfo timeZone,
+            IAnalyticsService analyticsService = null)
+        {
+            try
+            {
+                return value == default(DateTimeOffset) ? value : TimeZoneInfo.ConvertTime(value, timeZone);
+            }
+            catch (ArgumentOutOfRangeException)
+            {
+                analyticsService?.UnrepresentableDateErrorWhenConvertingDateToFormattedString?.Track(
```
Comment (kornelijepetak) [inline Toggl.Core.UI/Transformations/DateTimeToFormattedString.cs]: I've only noticed this right now.
I'd also rather have the app crash if `analyticsService` is null. If we don't have core functionality, the app should not work at all. I'd ditch the `?.` operator.

If we go with what you wrote here, what will happen is that an error will happen, and we will not even know it.

**Pull request merged**